analyzers/tests/SonarAnalyzer.UnitTest/TestCases/TypeMemberVisibility.cs

```diff
@@ -0,0 +1,182 @@
+//          ^^^^^^ Secondary
+            protected internal int ProtectedInternalProperty { get; } // Compliant: can be used in `InternalsVisibleTo` assemblies
+            internal int InternalProperty { get; } // Compliant: can be used in `InternalsVisibleTo` assemblies
+            protected int ProtectedProperty { get; } // Compliant: can be used in derived type
+            private int PrivateProperty { get; }
+        }
+
+        private class DerivedNestedPrivateClass : NestedPrivateClass
+        {
+            internal DerivedNestedPrivateClass()
+            {
+                var d = ProtectedProperty; // Protected properties can be accessed even if the base class is private.
+            }
+        }
+
+        protected class NestedProtectedClass
+        {
+            public int PropertyA { get; }
+//          ^^^^^^ Secondary
+            internal int PropertyB { get; }
+            protected internal int PropertyC { get; }
+            protected int PropertyD { get; }
+            private int PropertyF { get; }
+        }
+
+        private void Method()
+        {
+            var nestedPrivate = new NestedPrivateClass();
+            var x = nestedPrivate.InternalProperty; // internal properties can be used from nested private classes
+        }
+    }
+
+    internal struct Struct // Noncompliant
+    {
+        public class MyClass { }
+//      ^^^^^^ Secondary
+
+        public struct MyStruct { }
+//      ^^^^^^ Secondary
+        public enum MyEnum { }
+//      ^^^^^^ Secondary
+    }
+
+    public class Compliant // Class visibility upgrade makes members compliant
+    {
+        public static decimal A = 3.14m;
+        protected internal static decimal B = 1m;
+        internal static decimal C = 1m;
+
+        public int PropertyA { get; }
+        protected internal int PropertyB { get; }
+        internal int PropertyC { get; }
+
+        public int GetA() => 1;
+        protected internal int GetB() => 1;
+        internal int GetC() => 1;
+
+        public event EventHandler ClickA;
+        public delegate void DelegateA(string str);
+        public int this[int index] => 1;
+
+        public Compliant(int a) {}
+        protected internal Compliant(int a, int b) {}
+        internal Compliant(int a, int b, int c) {}
+
+        public struct NestedStructA
+        {
+            public bool FlipCoin() => false;
+        }
+
+        protected internal struct NestedStructB
+        {
+            internal bool FlipCoin() => false; // struct cannot have protected members
+        }
+
+        internal struct NestedStructC
+        {
+            internal bool FlipCoin() => false;
+        }
+
+        public class MyClass { }
+
+        public struct MyStruct { }
+
+        public enum MyEnum { }
+    }
+
+    internal sealed class UnsafeBinder : SerializationBinder
+    {
+        public override Type BindToType(string assemblyName, string typeName) => null; // Compliant - the overriden method is public
+    }
```

Comment:
Could you please add the following case?
```
    public interface ISomeInterface
    {
        void DoSomething();
    }

    internal class SomeClass: ISomeInterface // FP?
    {
        public void DoSomething() // FP
        {
            throw new NotImplementedException();
        }
    }
```
I would say that the case is a FP, do you agree with it?

Reply:
Nice find!